proposals/route-blinding.md

```diff
@@ -0,0 +1,242 @@
+extended with additional data in the future. It uses ChaCha20-Poly1305 as AEAD scheme.
+
+1. type: 10 (`enctlv`)
+2. data:
+    * [`...*byte`:`enctlv`]
+
+Once decrypted, the content of this encrypted blob is itself a TLV stream that may contain any
+tlv record defined in Bolt 4 (onion TLV namespace).
+
+### Creating a blinded route
+
+`N(r)` performs the following steps to create a blinded route:
+
+```text
+Initialization:
+
+  e(0) <- {0;1}^256
+  E(0) = e(0) * G
+
+Blinding:
+
+  For i = 0 to r-1:
+    ss(i) = H(e(i) * P(i)) = H(k(i) * E(i))         // shared secret known only by N(r) and N(i)
+    B(i) = HMAC256("blinded_node_id", ss(i)) * P(i) // Blinded node_id for N(i), private key known only by N(i)
+    rho(i) = HMAC256("rho", ss(i))                  // Key used to encrypt payload for N(i) by N(r)
+    e(i+1) = H(E(i) || ss(i)) * e(i)                // Ephemeral private key, only known by N(r)
+    E(i+1) = H(E(i) || ss(i)) * E(i)                // NB: N(i) must not learn e(i)
+
+Blinded route:
+  
+  (P(0),fees(0),cltv(0),encrypted_blob(0))
+  (B(1),fees(1),cltv(1),encrypted_blob(1))
+  ...
+  (B(r-1),fees(r-1),cltv(r-1),encrypted_blob(r-1))
+```
+
+Note that this is exactly the same construction as Sphinx, but at each hop we use the shared secret
+to derive a blinded `node_id` for `N(i)` for which the private key will only be known by `N(i)`.
+
+The recipient needs to provide `E(0)` and the blinded route to potential senders.
+The `encrypted_blob(i)` is encrypted with ChaCha20-Poly1305 using the `rho(i)` key, and contains
+the real `short_channel_id` to forward to (and potentially other fields). `E(i)` is included as
+additional authenticated data to detect probing attempts by the sender.
+
+Note that the introduction point uses the real `node_id`, not the blinded one, because the sender
+needs to be able to locate this introduction point and find a route to it. But the sender will send
+`E(0)`, which will allow the introduction point to compute the shared secret and correctly forward.
+
+Note that in the specific case of payments, the recipient can sign the invoice with `e(0)`.
+The sender will recover `E(0)` from the signature so no extra field needs to be added to Bolt 11.
+And this ensures the recipient doesn't reveal his real `node_id` through the invoice signature.
+
+### Sending to a blinded route
+
+The sender finds a route to the introduction point `N(0)`, and extends it with the blinded route.
+It then creates an onion for that route, and includes `E(0)` and `encrypted_blob(0)` in the onion
+payload for `N(0)`.
+
+When `N(0)` receives the onion and decrypts it, it finds `E(0)` in the payload and is able to
+compute the following:
+
+```text
+  ss(0) = H(k(0) * E(0))
+  rho(0) = HMAC256("rho", ss(0))
+  E(1) = H(E(0) || ss(0)) * E(0)
+```
+
+It uses `rho(0)` to decrypt the `encrypted_blob(0)` and discover the `scid` to forward to.
+It forwards the onion to the next node and includes `E(1)` in a TLV field in the message
```

Comment:
I don't get this part, where are you supposed to include E(1) ? In the `public_key` filed at the top-level packet structure? Any other location wouldn't break HMAC computed by original sender ?

Reply:
In the TLV extension of the `update_add_htlc` message. That's where it may create an issue for the anonymity set, because it means the length of this `update_add_htlc` message would be different from the length of "normal" `update_add_htlc` message which I don't like...

Reply:
Okay I'm not a fan of adding more fields in `update_add_hlc` because you may have third-party channel closure. Like Alice sends inside her packet a crappy `E(i)` to be processed between Caroll and Dave. At reception, Dave is going to close C-D link due to error attribution to Caroll.

It we spec this carefully that should be good but...

Reply:
> Dave is going to close C-D link due to error attribution to Caroll.

Why? This just leads to an onion decryption error, so Dave should fail that incoming HTLC but there's no reason to fail the channel. Here each node calculates the `E(i)` to send to the next node, so they can all make sure that it at least has the right format (so it won't trigger a parsing error which would lead to a close).